ENH Read mutiple excel sheets in single API call

jnmclarty opened the pull request:
Enables reading of multiple excel sheets in a single API call, reducing read time substantially. 

Essentially, 2O(n) becomes O(1) + O(n).
## Before

``` python
dfs = {}
for sheet in [‘Sheet1’,’Sheet2’,’Sheet3’]:
    #We have to open the file 3 times = Super Slow
    dfs[sheet] = pd.read_excel("TestData.xlsx",sheetname=sheet)
```
## This PR

``` python
#We open the file only 1 time = Much Faster, and it's just one python line
dfs = pd.read_excel("TestData.xlsx",sheetname=[‘Sheet1’,’Sheet2’,’Sheet3’])
```

...but as a bonus...

``` python
#the above is the same as...
dfs = pd.read_excel_sheets("TestData.xlsx",[‘Sheet1’,’Sheet2’,’Sheet3’])
#or, assuming those are there are only the 3 sheets...
dfs = pd.read_excel_sheets("TestData.xlsx")
#and this also works. But the dictionary returned has the integer keys instead of strings.
dfs = pd.read_excel_sheets("TestData.xlsx",[0,1,2])
```
## Return Objects

For the sheetname argument, specifying an int or a string, gets you a DataFrame (same as before the PR).  But, specify a list of strings/int, returns a dictionary of dataframes.  Specify None, and get all sheets in dictionary of dataframes, where the keys are the sheetnames.
## More Discussion

The first commit (2db986d) implements 100% of the functionality, while the next two commits (5c2304c and 1a53b01) layer on optional changes to the documentation and API (of the two, only one is actually mandatory, but keeping both is the most explicit).  Since the argument, sheetname, in read_excel was not plural, and defaults to 0, to maintain backwards compatibility AND create the “read all sheets” functionality associated with the non-default of None, the documentation of the argument gets really awkward.  Hence, creating read_excel_sheets with an argument sheetnames which defaults to None instead of 0.

Things I need feedback on:
- [x] Confirm print statement acceptability
- [x] Confirm if sheetname argument documentation should be changed (or not, to avoid confusion of dictionary-based return cases)
- [x] Confirm if read_excel_sheets addition to API is right approach

Things I still have to do:
- [x] Test/write tests
- [x] Update the release notes
- [x] Build the docs and check

Note: There might have been another way of doing this, but I couldn't easily figure it out, (SO/google wasn't helpful).  I just wanted to read my 623 sheets without it taking hours.

jnmclarty committed:
ENH read multiple sheets in read_excel()

shoyer commented:
Does it make sense to return a panel instead of a dict of dicts? Just something to think about...

jorisvandenbossche commented:
@shoyer I think it should at least be an option? Personally I would rather get a dict than a Panel, but that is more because I never worked with Panels before. Plus, there is no requirement that there are common indices in the different sheets, so a Panel is not really desired in that case I think?

@jnmclarty I think this is nice new functionality! To your questions:

- I wouldn't put the print statement there as default. If we want it, the user should at least be able to silence it (something like `verbose` argument, I don't know it something like that is already used somewhere else in pandas).
- For the `sheetname` documentation, I think the additions are certainly good. Only the last part (the different options, will have to be formatted a bit so it displays well, I will put some inline comments).
- For me, I don't think this new functionality warrants a new top-level function like `read_excel_sheets`. As doing `read_excel(.., sheetname=None)` is rather easy. But lets see what others think of this.  
  In any case, it could maybe be more prominently documented how to do this. Eg something like ```By default, it reads the first sheet. Specify the `sheetname` argument to select a specific sheet or use `sheetname=None` to read all sheets.``` that could be on top of the docstring (after the first line).

jorisvandenbossche commented on pandas/io/excel.py:
```diff
@@ -75,8 +74,19 @@ def read_excel(io, sheetname=0, **kwds):
+        where the keys match the sheet specified.  Set to None to get
+        all sheets, and use sheetnames as the dictionary keys.
+
+        Undefined -> Defaults to 0 -> 1st sheet as a DataFrame
```
You will have to use bullet points here to have this formatted nicely in the html docs.

jnmclarty commented:
@shoyer Thought about it, and my logic was the same as @jorisvandenbossche, plus, I'd wager that there are many times when the sheetnames are not something where an index makes sense.  Eg. "Sheet3", "Sheet4", "Sheet5" have data, but Sheet1 and Sheet2 got used for aggregation/reporting and were renamed properly.  Also, returning a dict forces the user to realize that the resulting dataframes are not ordered.  And...final point...returning a panel, we would have to make an assumption about the order for the sheetname=None case.

@jorisvandenbossche Thanks for the input.  I'll give this another 32 hours or so to let other feedback trickle in, and then begin implementing your comments.

jreback commented:
this should for sure be a ``dict of DataFrames``. Their is no reason to assume common indexes here. And if one wants it, then ``Panel(pd.read_excel(...,sheetnames=[.....]))`` would work (and you can add that to the docs in io.rst, need an example there as well).

agree with @jorisvandenbossche 
only a single API, namely ``read_excel(..)`` should exist here. ``sheetname`` taking a list is good.

shoyer commented:
Agreed, a dict makes more sense than a panel :).

jnmclarty commented:
Pretty much done, unless anybody else has feedback, on testing or the sphinx renderings below.  
![read_excel](https://cloud.githubusercontent.com/assets/6062071/6198578/83ff8be6-b3d9-11e4-8187-ff541691692d.png)
![io excel](https://cloud.githubusercontent.com/assets/6062071/6198579/8840f8fc-b3d9-11e4-9910-d78c9919a916.png)

jnmclarty commented:
Rebased.  Ready to merge. (Assuming Travis agrees)

jreback commented on doc/source/io.rst:
```diff
@@ -1949,56 +1949,100 @@ module and use the same parsing code as the above to convert tabular data into
+Specifying Sheets    
++++++++++++++++++
 
+- The second argument is ``sheetname``, not to be confused with ``ExcelFile.sheet_names``
```
I would move these first 2 points to a separate ``note::`` box.

jnmclarty replied:
Done, commit on it's way.

jreback commented on doc/source/io.rst:
```diff
@@ -1949,56 +1949,100 @@ module and use the same parsing code as the above to convert tabular data into
+
+.. code-block:: python
+
+   # If more than one sheet exists, returns a dictionary of DataFrames,
```
this breaks back-compat? e.g. the default should be to return a single-dateframe (e.g. ``sheetname=0``) is how the current interface works, right? shoulnd't ``sheetname=None`` return the dict of Frames?

jnmclarty replied:
No it doesn't.  ```sheetname``` still defaults to 0, and always returns one DataFrame, which is the first sheet.  If you set ```sheetname=None```, explicitly, it will either return the DateFrame if a single sheet exists, or a dictionary if multiple sheets exist.

jreback replied:
ahh I understand. 

However if you supply a ``sheetname=None|list`` then I think it *should* always return a dict. Its simply more consistent. Then you don't have a different return type if its a single sheet vs multiple sheets.

jnmclarty replied:
Are you sure?  If you don't know the sheetname, then you need to write the code to get the sheetname.  I know it's just sheetnames = mysheets.keys()[0], but...it's necessary.  I thought, it'd be more consistent.  1 sheet == DataFrame.  >1 sheet == dict.

jnmclarty replied:
Right now, even if you give it a single-element list, you get a dataframe back; consistent with the >1 Sheets == dict.

jreback replied:
@jorisvandenbossche @shoyer @TomAugspurger 

any thoughts on this?

jnmclarty replied:
I'm talking myself into going either way on this, I can see the merit of both. 

#1 One Sheet -> DataFrame; More Sheets -> Dict 
#2 ```str|int``` -> DataFrame; ```None|list``` -> Dict 

My next commit will switch #1 for #2.

jnmclarty replied:
Changed to # 2 ( @jreback 's thoughts) fabedcb

jreback commented on doc/source/whatsnew/v0.16.0.txt:
```diff
@@ -22,6 +22,10 @@ New features
 
+- The ``read_excel()`` function's sheetname argument now accepts a list and None, to get 
+  multiple or all sheets respectively.  If used, a dictionary is returned.
+- A ``verbose`` argument has been added to ``io.read_excel()``, defaults to False.
```
- explain the purpose of ``verbose``.
- add the issue number to this
- add a link to the documentation
- I think add a mini-example of using the new functionaility would be nice here (can be the same as in the docs).

jnmclarty replied:
[x] explain the purpose of verbose. (Done, need to confirm links with sphinx build)
[ ] add the issue number to this. (there was no issue number, so?)
[x] add a link to the documentation (Done, need to confirm links with sphinx build)
[x] mini-example (done, copied + pasted)
...commit coming.

jreback commented on pandas/io/excel.py:
```diff
@@ -296,76 +316,96 @@ def _parse_excel(self, sheetname=0, header=0, skiprows=None, skip_footer=0,
+                        should_parse[j] = self._should_parse(j, parse_cols)
+    
+                    if parse_cols is None or should_parse[j]:
+                        if typ == XL_CELL_DATE:
```
I would put all of this in to a locally defined function to make this code a bit more clean, e.g. something like ``parse_cell`` (you can define above)

jnmclarty replied:
Done...commit coming.

jreback commented on pandas/io/tests/test_excel.py:
```diff
@@ -423,7 +424,26 @@ def test_reader_converters(self):
+    
+    def test_reading_all_sheets(self):
+        # Test reading all sheetnames by setting sheetname to None,
+        # Ensure a dict is returned.
```
add the issue number as a comment on these tests

jnmclarty replied:
There was no issue number.  I just wanted the functionality, for myself.

jreback replied:
add the PR number (its the issue number when you don't have an issue); so it can be easily referenced.

jnmclarty replied:
Done. c20bb4ef

jreback commented on pandas/io/tests/test_excel.py:
```diff
@@ -423,7 +424,26 @@ def test_reader_converters(self):
+        dfs = read_excel(self.multisheet,sheetname=expected_keys)
+        tm.assert_contains_all(expected_keys,dfs.keys())
+    
     def test_reader_seconds(self):
```
can you add a test where you create some frames, write them to multiple-sheets, then read them in and compare.

jnmclarty replied:
While I can, and will, nothing (other than putting it in a for-loop) about the reading function changed.  As long as other tests do it, than any test I created would be redundant.

jreback replied:
I would rather have this tests, than relying on external sheets. Which maybe subtly different and non-reproducible. A complete end-to-end is simpler / more useful.

jnmclarty replied:
K, coming up.

jreback commented:
@jnmclarty I am a bit -1 on accepting mixed int/str for ``sheetname``. You can do it provided you account for the possibility of ambiguity (e.g. if I have a sheet named '1', but its the only sheet, what does '1' return? or 1 return?). IOW, we accept an int by default and seek to match it (if it exists).

So you would have to account for duplicates, and raise on ambiguity.

I would maybe suggest making the user specify either all ints (e.g. position based), or all names (e.g. by name).

jnmclarty commented:
It's currently just as ambiguous as it's always been.  You could have made the same comment, about ```'1'``` vs ```1``` before, the fact that the object is now in a list, is indifferent.

The moment we add exclusivity to the object type, THAT's when ```'1'``` will get ambiguous.

If you pass ```['1',1]``` (honestly, who would do this?), it will return a dict of DataFrames representing a sheet named "1", and the 2nd sheet if there is one. 

Let's say I've got a sheet named "MetaData", in my excel file, and I know I want the 5th sheet, but don't know the name. With my PR, you'd just go ['MetaData', 4]. Prohibiting mixing, you'd have no way of doing this (without more code)

jreback commented:
@jnmclarty ok as long as the behavior is well-defined and obvious. With a single sheetname it is either positional (if the name doesn't match), or is there (or a KeyError)?

with multiples then you can get dups and/or ambiguity. Maybe add a couple of tests to cement the behavior (ok with accepting mixed if its useful).

jnmclarty commented:
> With a single sheetname it is either positional (if the name doesn't match), or is there (or a KeyError)?

The bahviour is the same as before, just inside the list.  ```[1]``` is positional.  ```['1']``` is name.
No KeyErrors.

> with multiples then you can get dups and/or ambiguity

Assume the excel file has a sheet named "1", in the 1st position.  I handled 2 of 3 duplicates that would result from ```['1',0,0]```, in my latest commit.  So, passing ```['1',0,0]``` would return the same object as ```['1',0]```.  At this point though, maybe the user wants a duplicate copy of the DataFrame?  Honestly, at this point, we've got an excel file with a ridiculously unfortunate naming convention, and a Pandas user that insists on a single line API call.  I say, at this point, give the user the duplicates they request.

jnmclarty commented:
Oh, and @jreback  my original tests were already written to ensure the mixed-mode int/strings.  See line 447 of test_excel.py.

jnmclarty commented:
Ok, @jreback  I think we're good, see 3556753 and a73645b.  I'll squash momentarily.

jreback commented:
@jnmclarty duplicates are ok, but a KeyError is still a problem

e.g. a file with 2 elements, what should [1,2] do? raise a KeyError.

also need a rebase.

jnmclarty commented:
Did the rebase.

Why would ```[1,2]``` raise a KeyError? Assuming the workbook has 3 sheets, it would return the 2nd and 3rd sheet.

jnmclarty commented:
Oh, I get what you're saying now.

jnmclarty commented:
FWIW, d8775d7 is appropriate, but technically changes previous behaviour.  ```read_excel(...,sheetname=7,...)```, in the past, would return ```IndexError: list index out of range```

jnmclarty commented:
Multiple ways to skin a cat.  Will revert to IndexError.

jnmclarty commented:
Done (Using IndexError 0eb9c34)

jreback commented on doc/source/whatsnew/v0.16.0.txt:
```diff
@@ -226,65 +229,19 @@ Bug Fixes
-
-
-
-
```
put back this space, its there on purpose. To prevent collissions when mutiple prs are put in.

jnmclarty replied:
Sorry.  Didn't know that.  Thought I was being helpful.  I explicitly added a bunch of spaces, but I can revert back to the original, then replay just my changes excluding my "space cleanup", if that would be more optimal.  I'm guessing it would be, there will be many who's PRs would got stale as a result of me doing this, right?

jnmclarty replied:
K, replaced the EXACT same spaces.  75bc4e0  Squashed.  Ready to go.

jreback commented on doc/source/io.rst:
```diff
@@ -1949,56 +1949,103 @@ module and use the same parsing code as the above to convert tabular data into
+
+.. versionadded:: 0.16
+``read_excel`` can read more than one sheet, by setting ``sheetname`` to either
+a list of sheet names, a list of sheet positions, or None to read all sheets.
```
put ``None`` in back-ticks

jreback commented on doc/source/io.rst:
```diff
@@ -1949,56 +1949,103 @@ module and use the same parsing code as the above to convert tabular data into
-
-Using the sheet name:
+- Pass a list of either strings or integers, to return a dictionary of specified sheets.
+- Pass a None, to return a dictionary of all available sheets.
```
use back-ticks on None

jreback commented on doc/source/io.rst:
```diff
@@ -1949,56 +1949,103 @@ module and use the same parsing code as the above to convert tabular data into
+   # If more than one sheet exists, returns a dictionary of DataFrames,
+   # otherwise returns a DataFrame
+   read_excel('path_to_file.xls',sheetname=None)
+
```
I think the comments in the io.rst need to be updated here? (to conform with option 2)

jreback commented on doc/source/whatsnew/v0.16.0.txt:
```diff
@@ -190,6 +190,15 @@ Enhancements
 - Added ``StringMethods.find()`` and ``rfind()`` which behave as the same as standard ``str`` (:issue:`9386`)
 
 - Added ``StringMethods.isnumeric`` and ``isdecimal`` which behave as the same as standard ``str`` (:issue:`9439`)
+- The ``read_excel()`` function's :ref:`sheetname <_io.specifying_sheets>` argument now accepts a list and None, to get 
```
back-ticks on None, don't put spaces the end of a statement, just continue it on (it will get wrapped automatically)

jreback commented on doc/source/whatsnew/v0.16.0.txt:
```diff
@@ -190,6 +190,15 @@ Enhancements
+   # Returns the 1st and 4th sheet, as a dictionary of DataFrames.
+   read_excel('path_to_file.xls',sheetname=['Sheet1',3]) 
+  
+- A ``verbose`` argument has been augmented in ``io.read_excel()``, defaults to False. Set to True to print sheet names as they are parsed. (:issue:`9450`)
```
``pd.read_excel()``

jreback commented on pandas/io/excel.py:
```diff
@@ -75,8 +75,26 @@ def read_excel(io, sheetname=0, **kwds):
+        * Undefined -> Defaults to 0 -> 1st sheet as a DataFrame
+        * 1 -> 2nd sheet as a DataFrame
+        * "Sheet1" -> 1st sheet as a DataFrame
+        * [0,1,"Sheet5"] -> 1st, 2nd & 5th sheet as a dictionary of DateFrames
```
DataFrames

jreback commented on pandas/io/excel.py:
```diff
@@ -185,8 +203,25 @@ def parse(self, sheetname=0, header=0, skiprows=None, skip_footer=0,
+            * 1 -> 2nd sheet as a DataFrame
+            * "Sheet1" -> 1st sheet as a DataFrame
+            * [0,1,"Sheet5"] -> 1st, 2nd & 5th sheet as a dictionary of DateFrames
+            * None -> All sheets as a dictionary of DataFrames
```
DataFrames

jreback commented on pandas/io/excel.py:
```diff
@@ -223,6 +258,9 @@ def parse(self, sheetname=0, header=0, skiprows=None, skip_footer=0,
         has_index_names : boolean, default False
             True if the cols defined in index_col have an index name and are
             not in the header
+        verbose = bool, default False
```
use a ``:`` here, and say, boolean rather than bool

jnmclarty commented:
Fixed the Dat*e*Frames, ``None``s and ``pd.``s, and reference to my first implementation, 62d13ec.  Got my verbosity stupidity in prev commit.

jnmclarty commented:
Thanks @jreback, you're a machine.

jreback commented:
@jnmclarty hahha....happen to be home and its cold out!

anyhow, lgtm.

@jorisvandenbossche 
@TomAugspurger 
@shoyer 
have a look

jnmclarty commented:
Fixed error in the Exception handling I added at the 11th hour of this PR.  Squashed.

jnmclarty commented:
The IndexError exception I added breaks a test.  I thought it was a very trivial change, I guess not. I'm wondering if different exceptions are raised depending on the modules, or versions of modules, installed.  I reverted back my addition of IndexErrors, in 1783636 , but left it commented out. I believe Travis will not complain, with this final squash.  Assuming it passes, this PR can be merged, but either I, or somebody else should look into the exceptions that are thrown, and the exceptions that should be thrown, for all the various cases, and versions.  To me, this is low-value add, as the user of any given stack, will get a specific exception, and they will handle it according to their stack.

jreback commented on pandas/io/tests/test_excel.py:
```diff
@@ -423,7 +425,62 @@ def test_reader_converters(self):
+        dfs = [tdf(s) for s in sheets]
+        dfs = dict(zip(sheets,dfs))
+        
+        w = pd.ExcelWriter(self.multisheet2)
```
use a temporary filepath here, see [like this](https://github.com/pydata/pandas/blob/master/pandas/io/tests/test_excel.py#L275); you cannot assume that you can write to the test dirctory (e.g. the code); travis lets you , but on a regular install you normally cant'.

jnmclarty commented:
Ok, @jreback this PR is back to a state where it does not change any behaviour associated with any Exceptions anywhere.  It merely extends the same behaviour that would have existed before this PR.  So, ```sheetname=3```, will do the same thing as ```sheetname=[2,3]```, if there is no 4th sheet.  I removed my confusion inducing commented out exceptions that got added half-way through this PR discussion.  All our discussion surrounding IndexError is mute. 

I also implemented ensure_clean() in the one test.

Acknowledge my latest, and I'll squash sometime within 5 minutes to 12h time.

jreback commented:
lgtm

jnmclarty commented:
Explicitly, unless anybody else has any objections, this is ready to be merged.

jreback commented:
just needs a look for @jorisvandenbossche / @shoyer

shoyer commented:
I don't use Excel for serious analytics (thank goodness!) so I don't have much to add here.

My main thought is that this might be a good time to clean up the `sheetname`/`sheet_name` inconsistency between `read_excel` and `to_excel`.

jnmclarty commented:
Can we merge this, and open an issue for the ```sheetname```/```sheet_name``` thing?  I feel like at the same time as doing that, the codebase could use some cleaning up from a self-documenting perspective, for variable naming.  IMO, a PR like that, which doesn't change functionality, would be better isolated from one that does (like this PR we just did here).

shoyer commented:
@jnmclarty sounds good to me.

jorisvandenbossche commented on doc/source/io.rst:
```diff
@@ -1949,56 +1949,102 @@ module and use the same parsing code as the above to convert tabular data into
+~~~~~~~~~~~~~~~~~~~
+
+.. versionadded:: 0.16
+``read_excel`` can read more than one sheet, by setting ``sheetname`` to either
```
You need to add a blank line between  `.. versionadded:: 0.16` and the explanation (otherwise sphinx gives warnings)

jorisvandenbossche replied:
and the same with the two occurences below

jorisvandenbossche commented on doc/source/io.rst:
```diff
@@ -1949,56 +1949,102 @@ module and use the same parsing code as the above to convert tabular data into
 
 .. code-block:: python
-
+   # Returns a DataFrame
```
There needs to be a blank line between `.. code-block` and the actual content of the code block (`# Returns ..`), otherwise that line gets interpreted as arguments to `code-block` directive, leading to errors.

jorisvandenbossche commented on pandas/io/excel.py:
```diff
@@ -75,8 +75,26 @@ def read_excel(io, sheetname=0, **kwds):
+               
+        Available Cases
+
+        * Undefined -> Defaults to 0 -> 1st sheet as a DataFrame
```
"Undefined" sounds a bit strange. I would just say "Defaults to 0"

jorisvandenbossche commented on pandas/io/excel.py:
```diff
@@ -75,8 +75,26 @@ def read_excel(io, sheetname=0, **kwds):
+        Specify None to get all sheets.
+        
+        str|int -> DataFrame is returned.
+        list|None -> Dict of DataFrames is returned, with keys representing sheets.
```
You should also update the `Returns` section of the docstring with this info

jnmclarty commented:
@jorisvandenbossche all fixed, see e977432 (which, after I rebase momentarily, will be an outdated commit)

jreback commented:
@jnmclarty very nice enhancement. thanks for this!

pls have a look at the built docs and verify that everything looks kosher. If not, pls submit a followup!

http://pandas-docs.github.io/pandas-docs-travis/ (usually takes an hour; these are built on the 2.7 build from travis)

jnmclarty commented:
@jreback it looks to me like there's a problem with the travis built docs.  Does not look like my PR is a result, but the [index](http://pandas-docs.github.io/pandas-docs-travis/genindex.html) only has a few pages in it.

jreback commented:
I don't think the index is built on the travis docs. Only the 'main' parts are built (e.g. API is also not built). These looks ok to me.
